src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -91,40 +92,100 @@ function (indicator_hg::IndicatorHennemannGassner)(u::AbstractArray{<:Any,4},
-    end
+    apply_smoothing!(u, mesh, equations, dg, alpha, alpha_tmp, cache)
+
   end
 
   return alpha
 end
 
 
+function apply_smoothing!(u::AbstractArray{<:Any, 4}, mesh::Union{TreeMesh{2}, P4estMesh{2}}, equations, dg::DGSEM, alpha, alpha_tmp, cache)
+
+  # Diffuse alpha values by setting each alpha to at least 50% of neighboring elements' alpha
+  # Copy alpha values such that smoothing is indpedenent of the element access order
+  alpha_tmp .= alpha
+
+  # Loop over interfaces
+  for interface in eachinterface(dg, cache)
+    # Get neighboring element ids
+    left  = cache.interfaces.neighbor_ids[1, interface]
+    right = cache.interfaces.neighbor_ids[2, interface]
+
+    # Apply smoothing
+    alpha[left]  = max(alpha_tmp[left],  0.5 * alpha_tmp[right], alpha[left])
+    alpha[right] = max(alpha_tmp[right], 0.5 * alpha_tmp[left],  alpha[right])
+  end
+
+  # Loop over L2 mortars
+  for mortar in eachmortar(dg, cache)
+    # Get neighboring element ids
+    lower = cache.mortars.neighbor_ids[1, mortar]
+    upper = cache.mortars.neighbor_ids[2, mortar]
+    large = cache.mortars.neighbor_ids[3, mortar]
+
+    # Apply smoothing
+    alpha[lower] = max(alpha_tmp[lower], 0.5 * alpha_tmp[large], alpha[lower])
+    alpha[upper] = max(alpha_tmp[upper], 0.5 * alpha_tmp[large], alpha[upper])
+    alpha[large] = max(alpha_tmp[large], 0.5 * alpha_tmp[lower], alpha[large])
+    alpha[large] = max(alpha_tmp[large], 0.5 * alpha_tmp[upper], alpha[large])
+  end
+  
+end
+
+
+function apply_smoothing!(u::AbstractArray{<:Any, 4}, mesh::StructuredMesh{2}, equations, dg::DGSEM, alpha, alpha_tmp, cache)
+
+  # Diffuse alpha values by setting each alpha to at least 50% of neighboring elements' alpha
+  # Copy alpha values such that smoothing is indpedenent of the element access order
+  alpha_tmp .= alpha
+
+  # Loop over elements, because there is no interface container
+  for element in eachelement(dg,cache)
+    # Get neighboring element ids
+    left  = cache.elements.left_neighbors[1, element]
+    right = cache.elements.right_neighbors[1, element]
+    up    = cache.elements.left_neighbors[2, element] 
+    down  = cache.elements.right_neighbors[2, element]
+
+    # Apply smoothing
+    alpha[left]     = max(alpha_tmp[left],    0.5 * alpha_tmp[element], alpha[left])
+    alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[left],    alpha[element])
+    
+    alpha[right]    = max(alpha_tmp[right],   0.5 * alpha_tmp[element], alpha[right])
+    alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[right],   alpha[element])
+
+    alpha[up]       = max(alpha_tmp[up],      0.5 * alpha_tmp[element], alpha[up])
+    alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[up],      alpha[element])
+
+    alpha[down]     = max(alpha_tmp[down],    0.5 * alpha_tmp[element], alpha[down])
+    alpha[element]  = max(alpha_tmp[element], 0.5 * alpha_tmp[down],    alpha[element])
```

Comment:
I think in the structured mesh you only ever need to do the smoothing in one direction e.g., to the "left" in x, y, and z. By definition, the smoothing only needs to be done between two elements, so one of them is always the "right" and hence this should get you covered (this should also hold for periodic meshes). If I am not mistaken, here you are doing twice the work, i.e., each interface, across which smoothing is applied, is visited twice: once from the "left", and once from the "right".

If I didn't miss anything and it is sufficient to only consider the smoothing in one direction, you should also revisit the decision to store the `right_neighbors`. If the smoothing was the only use case, it should not be necessary anymore.

In the end, this should be easy to test by just commenting half the lines here and see if the results remain identical.

Reply:
This of course applies to 1D and 3D as well.

Reply:
Yes, you're right. I fixed this in 8716ee7.